### PR TITLE
Implement the Top-Level Cypher in the new translation layer.

### LINKED ISSUE #487: Fields relational/cypher fields are not resolved for custom queries/mutations returning Interfaces or Unions
**Describe the bug**
Related nodes, cypher fields are not resolved on interface/union types returned by custom cypher fields/mutations.  Queries/Mutations returning concrete types *do* correctly have requested fields and edges traversed for use in the response.

**Type definitions**
Given a simple schema:

```

type Author {
    name: String!  @relationship(type: "WROTE", direction: OUT)
}

type Director {
    name: String! @relationship(type: "DIRECTED", direction: OUT)
}

interface ThingWithTitleInterface {
    title: String!
}

type Book implements ThingWithTitleInterface {
   title: String!

    author: Author! @relationship(type: "WROTE", direction: IN)
    numberOfPages: Int!
}

type Movie implements ThingWithTitleInterface {
    title: String!

    director: Director! @relationship(type: "DIRECTED", direction: IN)
    runtime: Int!
}

union ThingWithTitle: Book | Movie

type Query {
    getByTitle(title: String!): [ThingWithTitle!] @cypher(statement: "<cypher query>")
    
    **** OR ****

    getByTitle(title: String!): [ThingWithTitleInterface!] @cypher(statement: "<cypher query>")
}
```

I should be able to call with a query like:

```
getByTitle(title: "Jurassic Park") {
    title
    ... on Book {
        author {
           name
       }
        numberOfPages
    }
    ... on Movie {
        director {
           name
       }
        runtime
    }
}
```

however, the call throws an error:

`Cannot return null for non-nullable field Movie.director`

Looking at the cypher being generated:

```
WITH apoc.cypher.runFirstColumn("<cypher query>", {auth: $auth, input: $input}, false) as x
                UNWIND x as this
                WITH this


RETURN this  AS this
```
It is clear that no attempt is being made to resolve connected edges, either director, or author, whichever concrete type is returned by `getByTitle`.

Fields and relationships *are* correctly evaluated *IFF* a concrete type is returned by the custom query or mutations, for example a custom query returning a specific Book or Movie.

**Expected behavior**
The GraphQL Library should correctly resolve and return data as requested by the caller.

GraphQL library is correctly populating the types, for example, Apollo GraphQL Sandbox correctly provides auto completion for fields provided by an interface, when the return type of the query is an interface.  If a custom query or mutation returns an interface or union type, the child fields should be correctly resolved *even if* that means a second call to Neo4j needs to be made based on the runtime resolved type.

At the very least, *interface* fields, including relationships, should be resolved correctly if an interface type is returned.

### PR DESCRIPTION
**Pull request by MacondoExpress**
Implement the Top-Level Cypher in the new translation layer.
Closes: #487

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4621.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 35 | 47 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 53 | 55 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 57 | 81 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 39 | 62 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2182 | 2219 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2102 | 2202 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 35 | 40 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 31 | 42 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2934 | 2732 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 47 | 138 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 95 | 89 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 117 | 164 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 50 | 60 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 214 | 176 | 3003 |
|  connections.Connection | 14049 | 14049 | 54 | 68 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 132 | 146 | 4516 |
|  create.SimpleMutation | 7 | 7 | 49 | 63 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 25 | 29 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 229 | 201 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 194 | 204 | 1183 |
|  2871.NestedRelationshipFilter | 23073 | 23086 | 91 | 63 | 4429 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 49 | 45 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 76 | 91 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 33 | 40 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 48 | 62 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 18 | 27 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 35 | 54 | 2174 |
|  query.QueryWhere | 9677 | 9665 | 38 | 43 | 2161 |
|  query.SimpleQueryWithNestedWhere | 9844 | 9832 | 48 | 52 | 2161 |
|  query.Nested | 10095502 | 10095502 | 6715 | 6836 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6666 | 6605 | 2004000 |
|  query.OrFilterOnRelationships | 42393 | 43057 | 139 | 177 | 1754 |
|  query.OrFilterOnRelationshipsAndNested | 36080 | 37172 | 186 | 240 | 1754 |
|  query.QueryWithNestedIn | 13452 | 13679 | 46 | 77 | 1245 |
|  query.NestedConnectionWhere | 9801 | 9801 | 54 | 68 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9850 | 10006 | 89 | 108 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18758 | 18962 | 130 | 158 | 1586 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 188 | 174 | 1134 |
|  query.Fulltext | 64 | 64 | 30 | 31 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 39 | 50 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 57 | 49 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 71 | 82 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14184 | 14174 | 57 | 63 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14184 | 14174 | 94 | 110 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14967 | 14659 | 97 | 123 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14493 | 14421 | 109 | 140 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 96 | 85 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14437 | 14443 | 117 | 114 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 39 | 53 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 84 | 92 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 71 | 84 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 54 | 67 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 61 | 65 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 300 | 259 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 233 | 299 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 96 | 115 | 2002 |

</details>
        
Old Schema Generation: 25.313s
Schema Generation: 23.643s
Old Subgraph Schema Generation: 28.900s
Subgraph Schema Generation: 28.315s

File: packages/graphql/src/translate/queryAST/ast/operations/CypherOperation.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        directed,
+        selection,
+    }: {
+        target: ConcreteEntityAdapter;
+        relationship?: RelationshipAdapter;
+        directed?: boolean;
+        selection: EntitySelection;
+    }) {
+        super({ target, relationship, directed, selection });
+    }
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/CypherOperation.ts]: Do we need this constructor? being the same as ReadOperation

File: packages/graphql/src/translate/queryAST/ast/selection/NodeSelection.ts [hunk 2]
```diff
@@ -27,23 +27,50 @@ export class NodeSelection extends EntitySelection {
...
         nestedContext: QueryASTContext<Cypher.Node>;
         selection: SelectionClause;
     } {
-        const node = createNodeFromEntity(this.target, context.neo4jGraphQLContext, this.alias);
-        const match = new Cypher.Match(node);
+        let node: Cypher.Node;
+        let matchPattern: Cypher.Pattern | undefined;
+
+        // useContextTarget is used when you have to select a node already matched,
+        // this could be simplified a lot, it's currently not possible as there is no way to build a Cypher.Node from an existing Cypher.Node.
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/selection/NodeSelection.ts:58]: I'm not sure this comment adds much tbh

File: packages/graphql/tests/integration/issues/487.int.test.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ import Neo4j from "../neo4j";
 import { Neo4jGraphQL } from "../../../src/classes";
 import { UniqueType } from "../../utils/graphql-types";
 
-describe("https://github.com/neo4j/graphql/issues/487", () => {
+describe("https://github.com/neo4j/graphql/issues/487, test on union", () => {
```
Comment (angrykoala) [inline packages/graphql/tests/integration/issues/487.int.test.ts]: (non-blocking) consider making 2 tests on the same describe, just to keep the same structure

**Review by angrykoala — DISMISSED** [review]

File: packages/graphql/src/translate/queryAST/ast/operations/CypherOperation.ts (hunk 1 above)
Comment (MacondoExpress) [reply packages/graphql/src/translate/queryAST/ast/operations/CypherOperation.ts]: Good catch!

**Pull request merged**